### ISSUE TITLE
Engine gets hung at some X%

### ISSUE THREAD

**Issue by g-karthik** [pytorch/ignite#828]
## ❓ Questions/Help/Support

I'm using an `Engine` for inference using a trained model and writing the inference outputs to a file. Specifically, I have something like

```
    model = # initialize some model
    output_writer = csv.writer(open(output_file_path, "w"), delimiter='\t') # initialize a writer

    def evaluate(engine, batch):
         example = batch[0]
         # do inference on example using model
         output_writer.writerow(["writing output of inference on batch to this writer"])
         return

    evaluator = Engine(evaluate)
    pbar = ProgressBar(persist=True)
    pbar.attach(evaluator)
    evaluator.run(data_loader) # here the data_loader is a default DataLoader with a batch size of 1
```

My first question is: is this approach the "right" way of doing what I want to do with ignite? Specifically, to do inference (beam search) on the model and write the output to a file?

My second question/concern is that in rare cases, I find this approach gets "hung" for some models I've trained. It gets hung at some X% (like 47% with my current model).

```
Epoch: [462/980]  47%|████▋      [01:51<01:43]
```

I did `watch -d -n 0.5 nvidia-smi` and I see that the process is alive, but the GPU usage is stuck at some Y% and not changing at all.

So then I killed this process and started it afresh, but this time, immediately after starting the process I did `strace -p <pid>` where `<pid>` is the process ID. I found that after it reached 47% and got stuck (it's weird that it is the same 47%), `strace` continues to output something but all of it is like the below:

```
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705347993}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705375070}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705402002}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705428796}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705475298}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705537752}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705595785}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705624538}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705653059}) = 0
clock_gettime(CLOCK_MONOTONIC_RAW, {107661, 705697886}) = 0
```

I am wondering if there is something wrong I'm doing -- but from looking at the strace output, it looks like the process is just hung and consuming `NO-OP` clock cycles?

@vfdev-5 any ideas on this?

**Comment by vfdev-5** [comment]
Hi @g-karthik 

> My first question is: is this approach the "right" way of doing what I want to do with ignite? 

Yes, this seems OK for me. In my project I do something similar but with batch size > 1 to accelerate NN stuff. Writing I do in a specfic handler attached on `Events.ITERATION_COMPLETED`, better factorization etc.

> My second question/concern is that in rare cases

yes, this looks strange. To debug that I would try the following things:
- dataloader `num_workers=0` = disable multiprocessing
- Is your dataloader is shuffled ? If not, try to check your 47-th element if it is correct and if you can do inference on that...
- try to comment output writing part

HTH

**Comment by g-karthik** [comment]
@vfdev-5 Can you share an example of how you write to a file with an ignite handler? I wanted to do the writing in a clean fashion by opening the file with the `with` command, but as you can see in my code snippet above, I didn't follow Pythonic best practice for file writing.

I figured out the reason the code is stuck at 47% -- there is a while loop in my decoding/inference which is not breaking, hence it is stuck.

**Comment by vfdev-5** [comment]
@g-karthik in my example it is about images and saving predictions as images. So, if it can help you, here is the code: 
- https://github.com/vfdev-5/UNOSAT_Challenge/blob/master/code/scripts/inference.py#L145
- https://github.com/vfdev-5/UNOSAT_Challenge/blob/37e12db083b794e1f135a6b7f7ab3df5a7bbff60/code/utils/handlers.py#L106
- https://github.com/vfdev-5/UNOSAT_Challenge/blob/37e12db083b794e1f135a6b7f7ab3df5a7bbff60/code/dataflow/io_utils.py#L211

If you write not that large text you can save it in RAM and save it once after the training.

> I figured out the reason the code is stuck at 47% -- there is a while loop in my decoding/inference which is not breaking, hence it is stuck.

That's a good news.

**Comment by vfdev-5** [comment]
@g-karthik did you resolved your issue ? Can we close this one ?

**Comment by g-karthik** [comment]
@vfdev-5 yes, I am closing it!
